drivers/watchdog/wdt_mcux_imx_wdog.c

```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ * Copyright (c) 2018, NXP
+
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_imx_wdog
+
+#include <drivers/watchdog.h>
+#include <fsl_wdog.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_mcux_wdog);
+
+#define WDOG_TMOUT_SEC(x)  ((int)(x * 2) - 1)
+
+struct mcux_wdog_config {
+	WDOG_Type *base;
+	void (*irq_config_func)(struct device *dev);
+};
+
+struct mcux_wdog_data {
+	wdt_callback_t callback;
+	wdog_config_t wdog_config;
+	bool timeout_valid;
+};
+
+static int mcux_wdog_setup(struct device *dev, uint8_t options)
+{
+	const struct mcux_wdog_config *config = dev->config;
+	struct mcux_wdog_data *data = dev->data;
+	WDOG_Type *base = config->base;
+
+	if (!data->timeout_valid) {
+		LOG_ERR("No valid timeouts installed");
+		return -EINVAL;
+	}
+
+	data->wdog_config.workMode.enableStop =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+
+	data->wdog_config.workMode.enableDebug =
+		(options & WDT_OPT_PAUSE_HALTED_BY_DBG) == 0U;
+
+	WDOG_Init(base, &data->wdog_config);
+	LOG_DBG("Setup the watchdog");
+
+	return 0;
+}
+
+static int mcux_wdog_disable(struct device *dev)
+{
+	const struct mcux_wdog_config *config = dev->config;
+	struct mcux_wdog_data *data = dev->data;
+	WDOG_Type *base = config->base;
+
+	WDOG_Deinit(base);
+	data->timeout_valid = false;
+	LOG_DBG("Disabled the watchdog");
+
+	return 0;
+}
+
+static int mcux_wdog_install_timeout(struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct mcux_wdog_data *data = dev->data;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+
+	WDOG_GetDefaultConfig(&data->wdog_config);
+
+	data->wdog_config.timeoutValue =
+		  WDOG_TMOUT_SEC(cfg->window.max / 1000.0f);
```

Comment:
Does the divisor need to be a float? Can it be an integer instead?

Reply:
NO, I think float is better, becasue our watchdog's resolution is 0.5(timeout periods from 0.5 to 128), but the register can only accept (0 - 127), so  i write a macro:#define WDOG_TMOUT_SEC(x)  ((int)(x * 2) - 1) that can accept a float. if the divisor is a integer, maybe we can not accept a float timeout value, such as 0.5(if we use integer divisor, the result maybe 500ms / 1000 = 0, has no effect), maybe a wasting of the resolution.

Reply:
If you multiply by two before the divide, you can achieve the same result in fixed point.

Reply:
Yes，You are right， I will rewrite the macro, and update this after

Reply:
Has updated，Please check